src/lib/slot.c

```diff
@@ -139,8 +144,134 @@ CK_RV slot_mechanism_list_get (CK_SLOT_ID slot_id, CK_MECHANISM_TYPE *mechanism_
+                break;
+            case CKM_EC_KEY_PAIR_GEN:
+                if (is_algorithm_supported(algs, TPM2_ALG_ECC)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_RSA_PKCS:
+                if (is_algorithm_supported(algs, TPM2_ALG_RSA)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_RSA_PKCS_KEY_PAIR_GEN:
+                if (is_algorithm_supported(algs, TPM2_ALG_RSA)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_RSA_PKCS_OAEP:
+                if (is_algorithm_supported(algs, TPM2_ALG_OAEP)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_RSA_X_509:
+                if (is_algorithm_supported(algs, TPM2_ALG_RSA)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+            case CKM_SHA_1:
+                if (is_algorithm_supported(algs, TPM2_ALG_SHA1)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_SHA1_RSA_PKCS:
+                if (is_algorithm_supported(algs, TPM2_ALG_SHA1) &&
+                    is_algorithm_supported(algs, TPM2_ALG_RSA)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_SHA256:
+                if (is_algorithm_supported(algs, TPM2_ALG_SHA256)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_SHA256_RSA_PKCS:
+                if (is_algorithm_supported(algs, TPM2_ALG_SHA256) &&
+                    is_algorithm_supported(algs, TPM2_ALG_RSA)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_SHA384:
+                if (is_algorithm_supported(algs, TPM2_ALG_SHA384)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_SHA384_RSA_PKCS:
+                if (is_algorithm_supported(algs, TPM2_ALG_SHA384) &&
+                    is_algorithm_supported(algs, TPM2_ALG_RSA)) {
+                    mechanism_list[supported] = mechs[i];
+                    supported++;
+                }
+                break;
+            case CKM_SHA512:
+                if (is_algorithm_supported(algs, TPM2_ALG_SHA512)) {
+                    mechanism_list[supported] = mechs[i];
```

Comment:
Interestingly, code coverage shows that the tpm simulator does not support sha512 - so sha512 is 'thoroughly tested' :P